packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -268,6 +271,15 @@ TextEditingValue _selectionAwareTextManipulation(
   TextSelection manipulatedSelection;
   if (selectionStartIndex < 0 || selectionEndIndex < 0) {
     manipulatedText = substringManipulation(value.text);
+  } else if (value.selection.isCollapsed) { // non-selection text manipulation
+    manipulatedText = substringManipulation(value.text);
+    // We only return the old valid value if the current value is not empty and
+    // if manipulation fails.
+    if (value.text.isNotEmpty && manipulatedText.isEmpty) {
+      manipulatedText = oldValue.text;
+    }
+    // move the carret to the end of the text
+    manipulatedSelection = TextSelection.collapsed(offset: manipulatedText.length);
```

Comment:
There is a bug that I believe is caused by this line:

  1. Type a bunch of numbers into the input.
  1. Move the cursor to the middle of the text.
  1. Insert another number at the new cursor position.
  1. The cursor jumps to the end of the line, but it should stay at its position.

Reply:
Noted. I must work on that soon

Reply:
Hey @justinmc I was unable to reproduce this issue, can you provide the code that you've used?

Reply:
I found it :)

Reply:
@justinmc I did some updates!